### PR TITLE
ROS Communication Testing: Add Tests for Topic Creation and Message Chain Verification

### LINKED ISSUE #55: Test that topics are created, verify the chain of message trough the nodes (scaler - AGeVaR - communication, etc.)

### PR DESCRIPTION
**Pull request by Yunus-Selim**
closes #55 

This code aims to test for the existence of topics and the existence of topic messages, to make sure there is communication between nodes. But it might have some bugs for the other config files. I pull request it for early analysis before the meeting.

And I have a few questions about the code:
1. Do you want me to use something else like correctness of messages for the assertion instead of checking the existence of topic messages?
2. Should I test for the **RobotStatePublisher** node for this task?
3. Should I complete the **TODO**s in the test code? Do you think that they are necessary?

### CODE REVIEW

**Comment by luca-byte** [comment]
1. Yes, but in separate test files, also we may want to check the frequency of the messages on the topic. There is a Sampling frequency in the remote controller (remote_controller) 80ms and one for the telemetry/feedback 100Hz (PicoLowLevel)
2. No
3. Yes, by using the data from YAML files, harrdcoded. 

Implement the Canbus in python to test feedback reception from the communication node. Also makes sense to test for failures

**Review by luca-byte — CHANGES_REQUESTED** [review]
Excellent work!! 💯 

There are some problems (namely the impossibility to run tests with colcon and the communication node tests) and I left some considerations about those. 

Moreover I think that the tests/classes/functions could benefit from having some docstrings/comments that explain briefly the test objective and the method used to test it.

I also kindly ask you to write a more detailed description of the approach (values selected for the test, how are oracles derived, ...) on Outline, in the section I just created.

File: reseq_ros2/test/test_node_communication.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+        problems = []
+        if len(self.msgs[:idx]) != self.end_effector_num:
+            problems.append(f"Number of topics doesn't match. Expected {self.end_effector_num}, got {len(self.msgs[:idx])}.")
+        for i in range(idx):
+            if not bool(self.msgs[i]):
+                problems.append(f"Couldn't get message from '/reseq/module{self.address + (i % (idx//3))}/end_effector/{['pitch', 'head_pitch', 'head_roll'][i // (idx//3)]}/setpoint' topic.")
+        self.assertTrue(len(problems) == 0, f"\n{''.join(problems)}")
+
+    def test_4_communication_topics(self):
```
[the remark this answers is not in the document]
Comment (Yunus-Selim) [reply reseq_ros2/test/test_node_communication.py:201]: Done! But I couldn't understand what caused the SocketcanBus error as I don't get this error when I use `launch_test <path_to_test_file>` I hope it's fixed after my second commit. Let me know if the problem persists.

[the remark this answers is not in the document]
Comment (luca-byte) [reply reseq_ros2/test/test_node_communication.py:216]: Yes it was not properly an error. Before the ...ok printed by test there was a log message (not properly shutdown)

**Comment by luca-byte** [comment]
I noticed that with the pytest mark the tests are visualized as only one test, thus making it difficult to see if one or more tests are failing.

Moreover I came to the conclusion that my suggestion to add the script was not a good idea for many reasons (simply because colcon does not allow to use the stdin, ...) so probably you will have to remove it (sorry 😁) however I still believe that we should skip tests when the interface is not available (and your approach to check the ip link output is the best onne in my opinion). 

As per the first point however, I would put the test that require the CAN interface in separate files, so that it is clear that they were skipped even when merged in one test by launch_test in colcon

Regarding the socketcan shutdown I intended shutting it down at the end of the test, since now (with the new implementation, a failure to setup it blocks also tests which are independent from it. I meant something like
```py
def test_with_canbus(...):
  # setup canbus
  ...
  canbus.shutdown(...)
```
or something similar, but if  you agree on separating the tests depending on the CAN we can do as you did in these tests.

**Comment by luca-byte** [comment]
Regarding documentation: 👌

**Review by luca-byte — COMMENTED** [review]
Great job 💯 

I noticed that between this PR and the ones testing the nodes, there is a lot of duplicated code (launch, ...). Would it be possible to merge together that code to improve readability and reduce duplicated code, maybe abstracting it in some classes/functions?

**Comment by luca-byte** [comment]
I do not like much this approach because we are polluting the production code with code which is never executed by the robot (unlike launch utilities). Isn't it possible to include the file from the same folder (if flake complains in the inclusion you can silence it using noqa on the offending line)

After that I will approve the PR. Great job 🎯

**Comment by Yunus-Selim** [comment]
It can be merged without problem now.

**Review by luca-byte — APPROVED** [review]
LGTM

**Pull request merged**